janusgraph-berkeleyje/src/main/java/org/janusgraph/diskstorage/berkeleyje/BerkeleyJEKeyValueStore.java

```diff
@@ -131,51 +130,65 @@ public void acquireLock(StaticBuffer key, StaticBuffer expectedValue, StoreTrans
         final StaticBuffer keyStart = query.getStart();
         final StaticBuffer keyEnd = query.getEnd();
         final KeySelector selector = query.getKeySelector();
-        final List<KeyValueEntry> result = new ArrayList<>();
         final DatabaseEntry foundKey = keyStart.as(ENTRY_FACTORY);
         final DatabaseEntry foundData = new DatabaseEntry();
-
         final Cursor cursor = openCursor(txh);
-        try {
-            OperationStatus status = cursor.getSearchKeyRange(foundKey, foundData, getLockMode(txh));
-            //Iterate until given condition is satisfied or end of records
-            while (status == OperationStatus.SUCCESS) {
-                StaticBuffer key = getBuffer(foundKey);
-
-                if (key.compareTo(keyEnd) >= 0)
-                    break;
-
-                if (selector.include(key)) {
-                    result.add(new KeyValueEntry(key, getBuffer(foundData)));
-                }
-
-                if (selector.reachedLimit())
-                    break;
-
-                status = cursor.getNext(foundKey, foundData, getLockMode(txh));
-            }
-        } catch (Exception e) {
-            throw new PermanentBackendException(e);
-        } finally {
-            closeCursor(txh, cursor);
-        }
-        log.trace("db={}, op=getSlice, tx={}, resultcount={}", name, txh, result.size());
 
         return new RecordIterator<KeyValueEntry>() {
-            private final Iterator<KeyValueEntry> entries = result.iterator();
+            private OperationStatus status;
+            private KeyValueEntry current;
 
             @Override
             public boolean hasNext() {
-                return entries.hasNext();
+                if (current == null) {
+                    current = getNextEntry();
+                }
+                return current != null;
             }
 
             @Override
             public KeyValueEntry next() {
-                return entries.next();
+                if (!hasNext()) {
+                    throw new NoSuchElementException();
+                }
+                KeyValueEntry next = current;
+                current = null;
+                return next;
+            }
+
+            private KeyValueEntry getNextEntry() {
+                if (status != null && status != OperationStatus.SUCCESS) {
+                    return null;
+                }
+                if (selector.reachedLimit()) {
+                    return null;
+                }
+                do {
```

Comment:
How about using a normal while loop with the condition `!selector.reachedLimit()`?

Reply:
Ok